# Task 1 - May Retention Rate Calculation 

In [1]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import plotly.graph_objects as go
import plotly.express as px
import plotly.io as pio
pio.renderers.default = 'iframe_connected'
%matplotlib inline

### 1. Reading and Exploration of Data

In [2]:
sales_data = pd.read_excel("dataset1.xlsx")
sales_data.head()

week  month creation_date  sales_order_id  retailer_id sales_order_status  \
0    14      4    2020-04-01               1        18564          Delivered   
1    14      4    2020-04-01               5        28140          Delivered   
2    14      4    2020-04-01               9        32552           Canceled   
3    14      4    2020-04-01              13        19299          Delivered   
4    14      4    2020-04-01              17         5901          Delivered   

       channel   district_ar  order_price_before_discount  discount  
0  Application     العمرانية                      2119.75      0.00  
1  Application      ميت عقبة                      2242.75      0.10  
2  Application  مدينة السلام                      1094.50      8.65  
3  Application      ميت عقبة                      3448.50     24.51  
4  Application       المطرية                       910.95      0.10

In [3]:
sales_data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21716 entries, 0 to 21715
Data columns (total 10 columns):
 #   Column                       Non-Null Count  Dtype  
---  ------                       --------------  -----  
 0   week                         21716 non-null  int64  
 1   month                        21716 non-null  int64  
 2   creation_date                21716 non-null  object 
 3   sales_order_id               21716 non-null  int64  
 4   retailer_id                  21716 non-null  int64  
 5   sales_order_status           21716 non-null  object 
 6   channel                      21716 non-null  object 
 7   district_ar                  21716 non-null  object 
 8   order_price_before_discount  21716 non-null  float64
 9   discount                     21716 non-null  float64
dtypes: float64(2), int64(4), object(4)
memory usage: 1.7+ MB


In [4]:
# Adding Revenue column
sales_data['revenue'] = sales_data['order_price_before_discount'] - sales_data['discount']
sales_data

week  month creation_date  sales_order_id  retailer_id  \
0        14      4    2020-04-01               1        18564   
1        14      4    2020-04-01               5        28140   
2        14      4    2020-04-01               9        32552   
3        14      4    2020-04-01              13        19299   
4        14      4    2020-04-01              17         5901   
...     ...    ...           ...             ...          ...   
21711    25      6    2020-06-15           86872        30883   
21712    25      6    2020-06-15           86876         5619   
21713    25      6    2020-06-15           86880        16638   
21714    25      6    2020-06-15           86884        36652   
21715    25      6    2020-06-15           86888        39303   

      sales_order_status      channel   district_ar  \
0              Delivered  Application     العمرانية   
1              Delivered  Application      ميت عقبة   
2               Canceled  Application  مدينة السلام   
3              Delivered  Application      ميت عقبة   
4              Delivered  Application       المطرية   
...                  ...          ...           ...   
21711          Delivered  Application         الدقي   
21712           Canceled  Application        الخصوص   
21713          Delivered  Application          فيصل   
21714          Delivered  Application         مسطرد   
21715          Delivered  Application    دار السلام   

       order_price_before_discount  discount  revenue  
0                          2119.75      0.00  2119.75  
1                          2242.75      0.10  2242.65  
2                          1094.50      8.65  1085.85  
3                          3448.50     24.51  3423.99  
4                           910.95      0.10   910.85  
...                            ...       ...      ...  
21711                      3309.00     46.52  3262.48  
21712                      2805.10     43.69  2761.41  
21713                      2294.20     39.76  2254.44  
21714                      1660.35     18.92  1641.43  
21715                       966.60      0.10   966.50  

[21716 rows x 11 columns]

### 2. Filtering April Data

In [5]:
april_data = sales_data[sales_data['month'] == 4]

# Calculating New retailers count in April
april_retailers = april_data['retailer_id'].unique().tolist()
april_retailers_count = len(april_retailers)
april_retailers_count

5004

#### Insights
- Till now we have the number of new retailers at the end of april month __(5004 New Retailer)__. 
- In order to calculate the retention rate for May, we need to know:
    - __How many customer of the 5004 retailers have continued to use MaxAB in May?__

In [6]:
# Getting Unique retailers in April
unique_april_retailers = april_data.drop_duplicates(subset=['retailer_id']).reset_index()

# Dividing retailers per Channel
april_app_user_ids = unique_april_retailers[unique_april_retailers['channel'] == 'Application']['retailer_id'].tolist()
april_salesagent_user_ids = unique_april_retailers[unique_april_retailers['channel'] == 'salesagent']['retailer_id'].tolist()
april_telesales_user_ids = unique_april_retailers[unique_april_retailers['channel'] == 'telesales']['retailer_id'].tolist()

### 3. Filtering May's Retained retailers Data

In [7]:
may_data = sales_data[sales_data['month'] == 5]

# check if retailer ID is in april_retailers
# If yes: Retained User Else: New Customer
pd.set_option('mode.chained_assignment', None)
may_data['is_retained'] = may_data[ 'retailer_id'].isin(april_retailers)

# Applying checks for each channel users
may_data['app_retained'] = may_data[ 'retailer_id'].isin(april_app_user_ids)
may_data['sales_agent_retained'] = may_data[ 'retailer_id'].isin(april_salesagent_user_ids)
may_data['telesales_retained'] = may_data[ 'retailer_id'].isin(april_telesales_user_ids)
may_data

week  month creation_date  sales_order_id  retailer_id  \
8661     18      5    2020-05-01           34647        26658   
8662     18      5    2020-05-01           34651         8913   
8663     18      5    2020-05-01           34655        20805   
8664     18      5    2020-05-01           34659        24977   
8665     18      5    2020-05-01           34663         8544   
...     ...    ...           ...             ...          ...   
16913    23      5    2020-05-31           67664        17123   
16914    23      5    2020-05-31           67668        30589   
16915    23      5    2020-05-31           67672        24375   
16916    23      5    2020-05-31           67676        13651   
16917    23      5    2020-05-31           67680         5374   

      sales_order_status      channel    district_ar  \
8661           Delivered  Application         المقطم   
8662           Delivered  Application        المطرية   
8663           Delivered  Application    باب الشعرية   
8664           Delivered  Application        عين شمس   
8665           Delivered  Application    حدائق القبة   
...                  ...          ...            ...   
16913          Delivered    telesales          هرم 2   
16914          Delivered  Application           فيصل   
16915          Delivered  Application        عين شمس   
16916          Delivered  Application  بولاق الدكرور   
16917          Delivered  Application         امبابة   

       order_price_before_discount  discount   revenue  is_retained  \
8661                       1359.20      3.59   1355.61        False   
8662                       3300.45     14.33   3286.12         True   
8663                      10947.15     39.51  10907.64         True   
8664                       3463.75     23.10   3440.65         True   
8665                       2981.75     29.07   2952.68         True   
...                            ...       ...       ...          ...   
16913                      1922.25     12.82   1909.43         True   
16914                      1367.00      0.00   1367.00         True   
16915                      2180.50     39.09   2141.41         True   
16916                      3097.20     50.74   3046.46         True   
16917                      3746.70     33.89   3712.81         True   

       app_retained  sales_agent_retained  telesales_retained  
8661          False                 False               False  
8662           True                 False               False  
8663           True                 False               False  
8664           True                 False               False  
8665           True                 False               False  
...             ...                   ...                 ...  
16913         False                 False                True  
16914          True                 False               False  
16915          True                 False               False  
16916          True                 False               False  
16917          True                 False               False  

[8257 rows x 15 columns]

In [8]:
may_data['is_retained'].value_counts()

True     4516
False    3741
Name: is_retained, dtype: int64

In [9]:
# Calculating the number of Retained retailers from April across different channels
retained_retailers_count = may_data['is_retained'].sum()
app_retained_retailers_count = may_data['app_retained'].sum()
sales_agent_retained_retailers_count = may_data['sales_agent_retained'].sum()
telesales_retained_retailers_count = may_data['telesales_retained'].sum()

## 

In [10]:
# Calculating May Retention Rate for every channel
may_retention_rate = np.round((retained_retailers_count / april_retailers_count)*100, 1)
app_may_retention_rate = np.round( app_retained_retailers_count / len(april_app_user_ids)*100, 1)
salesagent_may_retention_rate = np.round(sales_agent_retained_retailers_count / len(april_salesagent_user_ids)*100, 1)
telesales_may_retention_rate = np.round(telesales_retained_retailers_count / len(april_telesales_user_ids)*100 ,1)

# Plot the results
fig =  px.bar(x=['All Users', 'Application', 'Sales Agent', 'Telesales'], 
              y=[may_retention_rate, app_may_retention_rate,
                 salesagent_may_retention_rate,
                 telesales_may_retention_rate],
              text=[may_retention_rate, app_may_retention_rate, salesagent_may_retention_rate, telesales_may_retention_rate])

fig.update_xaxes(ticklabelposition="inside top", title='Channel')
fig.update_yaxes(ticklabelposition="inside top", title='Retention Rate')
fig.update_layout(font=dict(family='Arial', size=20))
fig.update_traces(marker_color=['#DC3912','#00B5F5','#00B5F5','#00B5F5'])
fig.show()

# Task 2 - Percentage Growth of May over April for every District

In [11]:
april_may_data = sales_data[sales_data['month'] != 6]

### 1. Number of Orders

In [12]:
number_of_orders = pd.pivot_table(data=april_may_data, 
                                  values='sales_order_id', 
                                  index='month', 
                                  columns='district_ar', 
                                  aggfunc='count' )

number_of_orders.dropna(axis=1, inplace=True)
number_of_orders

district_ar  أرض اللواء  البراجيل  الجبل الاصفر  الخانكة  الخصوص  الدقي  \
month                                                                     
4                 205.0      79.0          54.0      9.0   116.0   93.0   
5                 189.0      75.0          47.0     14.0   140.0   77.0   

district_ar  الزاوية الحمراء  الزيتون  السبتية  الشرابية  ...  مدينة الحرفيين  \
month                                                     ...                   
4                      137.0    173.0     68.0     130.0  ...           102.0   
5                      116.0    148.0     68.0      91.0  ...           105.0   

district_ar  مدينة السلام  مساكن شيراتون  مسطرد  منطي  ميت حلفا  ميت عقبة  \
month                                                                       
4                   122.0           15.0   87.0  21.0      10.0      62.0   
5                    86.0            5.0   70.0  19.0       6.0      71.0   

district_ar  ميت نما  هرم 2  هرم 3  
month                               
4                9.0  292.0  267.0  
5                9.0  270.0  209.0  

[2 rows x 64 columns]

In [13]:
# Order Growth Rate = (# of Orders in May - # of Orders in April) / # of Orders in April
temp_df = pd.DataFrame(index=number_of_orders.columns,
                       columns=['number_of_orders_growth'])
temp_df['number_of_orders_growth'] = np.round(((number_of_orders.loc[5,:] - number_of_orders.loc[4,:]) / number_of_orders.loc[4,:])*100,1)
orders = temp_df.sort_values('number_of_orders_growth', ascending=False)
orders.head()

number_of_orders_growth
district_ar                          
شبين القناطر                    200.0
القطاوي                         100.0
بلقس                             88.9
قلج                              77.8
الخانكة                          55.6

In [14]:
# Plot
data = px.treemap(data_frame= temp_df, 
                  path=[temp_df.index, 'number_of_orders_growth'], 
                  color_continuous_scale='deep',
                  color='number_of_orders_growth')

fig = go.Figure(data=data)      
fig.update_layout(font=dict(family='Arial', size=12), 
                  width=1200, 
                  height=800)
fig.show()

In [15]:
import plotly.express as px
from textwrap import wrap

named_colorscales = px.colors.named_colorscales()
print("\n".join(wrap("".join('{:<12}'.format(c) for c in named_colorscales), 96)))

aggrnyl     agsunset    blackbody   bluered     blues       blugrn      bluyl       brwnyl
bugn        bupu        burg        burgyl      cividis     darkmint    electric    emrld
gnbu        greens      greys       hot         inferno     jet         magenta     magma
mint        orrd        oranges     oryel       peach       pinkyl      plasma      plotly3
pubu        pubugn      purd        purp        purples     purpor      rainbow     rdbu
rdpu        redor       reds        sunset      sunsetdark  teal        tealgrn     turbo
viridis     ylgn        ylgnbu      ylorbr      ylorrd      algae       amp         deep
dense       gray        haline      ice         matter      solar       speed       tempo
thermal     turbid      armyrose    brbg        earth       fall        geyser      prgn
piyg        picnic      portland    puor        rdgy        rdylbu      rdylgn      spectral
tealrose    temps       tropic      balance     curl        delta       oxy         edge
hsv     

In [22]:
print(px.colors.sequential.Aggrnyl)

['rgb(36, 86, 104)', 'rgb(15, 114, 121)', 'rgb(13, 143, 129)', 'rgb(57, 171, 126)', 'rgb(110, 197, 116)', 'rgb(169, 220, 103)', 'rgb(237, 239, 93)']


### 2. Distinct Retailers Per Distrcit

In [15]:
distinct_retailers = pd.pivot_table(data=april_may_data, 
                                    values='retailer_id', 
                                    index='month', 
                                    columns='district_ar', 
                                    aggfunc=pd.Series.nunique)

distinct_retailers.dropna(axis=1, inplace=True)
distinct_retailers

district_ar  أرض اللواء  البراجيل  الجبل الاصفر  الخانكة  الخصوص  الدقي  \
month                                                                     
4                 115.0      53.0          31.0      3.0    76.0   47.0   
5                 121.0      50.0          32.0      6.0    93.0   43.0   

district_ar  الزاوية الحمراء  الزيتون  السبتية  الشرابية  ...  مدينة الحرفيين  \
month                                                     ...                   
4                       83.0     99.0     36.0      84.0  ...            59.0   
5                       70.0     87.0     39.0      57.0  ...            64.0   

district_ar  مدينة السلام  مساكن شيراتون  مسطرد  منطي  ميت حلفا  ميت عقبة  \
month                                                                       
4                    75.0            8.0   48.0   9.0       5.0      31.0   
5                    57.0            5.0   49.0  16.0       5.0      45.0   

district_ar  ميت نما  هرم 2  هرم 3  
month                               
4                6.0  182.0  145.0  
5                5.0  178.0  129.0  

[2 rows x 64 columns]

In [16]:
# Distinct Retailers Growth = (# of Distinct Retailers in May - # of Distinct Retailers in April) / # of Distinct Retailers in April
temp_df['distinct_retailers_growth'] = np.round(((distinct_retailers.loc[5,:] - distinct_retailers.loc[4,:]) / distinct_retailers.loc[4,:])*100,1)
retailers = temp_df.sort_values('distinct_retailers_growth', ascending=False)
retailers.head()

number_of_orders_growth  distinct_retailers_growth
district_ar                                                    
القطاوي                        100.0                      100.0
قلج                             77.8                      100.0
الخانكة                         55.6                      100.0
منطي                            -9.5                       77.8
بشتيل                           17.6                       60.3

In [17]:
data = px.treemap(data_frame= temp_df, 
                  path=[temp_df.index, 'distinct_retailers_growth'], 
                  color_continuous_scale='deep',
                  color='distinct_retailers_growth')

fig = go.Figure(data=data)
fig.update_layout(font=dict(family='Arial', size=12), 
                  width=1200, 
                  height=800)
fig.show()

### 3. Average Order Price Per District

In [18]:
average_order_price = pd.pivot_table(april_may_data, 
                                         values='order_price_before_discount', 
                                         index='month', 
                                         columns='district_ar' , 
                                         aggfunc='mean' 
                                         )
average_order_price.dropna(axis=1, inplace=True)
average_order_price

district_ar   أرض اللواء     البراجيل  الجبل الاصفر      الخانكة       الخصوص  \
month                                                                           
4            3272.263618  2876.951899   3477.589352  7651.233333  3126.856695   
5            3037.944180  2424.870667   2960.794681  7206.217143  2622.416071   

district_ar        الدقي  الزاوية الحمراء      الزيتون      السبتية  \
month                                                                 
4            2695.287688      3357.535182  3243.743988  2952.915221   
5            2165.114286      2133.535431  3068.229932  3161.604412   

district_ar     الشرابية  ...  مدينة الحرفيين  مدينة السلام  مساكن شيراتون  \
month                     ...                                                
4            2970.427962  ...     3982.986912   3069.693989    3127.032667   
5            2551.092088  ...     2586.952190   2312.982326    1583.980000   

district_ar        مسطرد         منطي     ميت حلفا     ميت عقبة      ميت نما  \
month                                                                          
4            3253.140057  2811.240476  4209.955000  2390.833065  2928.800000   
5            3093.444571  2657.944737  2037.516667  2156.069718  2051.998889   

district_ar        هرم 2        هرم 3  
month                                  
4            2664.678733  3022.841167  
5            2584.597852  3091.282105  

[2 rows x 64 columns]

In [19]:
# Average Order Price Growth = (Average Order Price in May - Average Order Price in April) / Average Order Price in April
temp_df['average_order_price_growth'] = np.round(((average_order_price.loc[5,:] - average_order_price.loc[4,:]) / average_order_price.loc[4,:])*100,1)
average_price = temp_df.sort_values('average_order_price_growth', ascending=False)
average_price

number_of_orders_growth  distinct_retailers_growth  \
district_ar                                                           
المطرية                            -19.0                       -4.4   
شبين القناطر                       200.0                        0.0   
القناطر الخيرية                     12.3                       -7.1   
شبرا مصر                             4.8                        0.0   
بهتيم                              -39.6                      -30.6   
...                                  ...                        ...   
الزاوية الحمراء                    -15.3                      -15.7   
مساكن شيراتون                      -66.7                      -37.5   
ميت حلفا                           -40.0                        0.0   
عمرانيه 3                          -75.0                      -63.6   
القطاوي                            100.0                      100.0   

                 average_order_price_growth  
district_ar                                  
المطرية                                13.3  
شبين القناطر                           10.4  
القناطر الخيرية                         9.3  
شبرا مصر                                7.9  
بهتيم                                   7.6  
...                                     ...  
الزاوية الحمراء                       -36.5  
مساكن شيراتون                         -49.3  
ميت حلفا                              -51.6  
عمرانيه 3                             -58.0  
القطاوي                               -72.3  

[64 rows x 3 columns]

In [20]:
data = px.treemap(data_frame= temp_df, 
                  path=[temp_df.index, 'average_order_price_growth'], 
                  color_continuous_scale='deep',
                  color='average_order_price_growth')

fig = go.Figure(data=data)
fig.update_layout(font=dict(family='Times New Roman', size=12), 
                  width=1200, 
                  height=800)
fig.show()

### 4. Total Revenue Growth Per District

In [21]:
total_monthly_revenue = pd.pivot_table(april_may_data, 
                                       values='order_price_before_discount', 
                                       index='month',
                                       columns='district_ar' , 
                                       aggfunc='sum')

total_monthly_revenue.dropna(axis=1, inplace=True)
total_monthly_revenue

district_ar     أرض اللواء  البراجيل  الجبل الاصفر    الخانكة         الخصوص  \
month                                                                          
4            670814.041667  227279.2    187789.825   68861.10  362715.376667   
5            574171.450000  181865.3    139157.350  100887.04  367138.250000   

district_ar       الدقي  الزاوية الحمراء    الزيتون     السبتية    الشرابية  \
month                                                                         
4            250661.755        459982.32  561167.71  200798.235  386155.635   
5            166713.800        247490.11  454098.03  214989.100  232149.380   

district_ar  ...  مدينة الحرفيين   مدينة السلام  مساكن شيراتون       مسطرد  \
month        ...                                                             
4            ...      406264.665  374502.666667       46905.49  283023.185   
5            ...      271629.980  198916.480000        7919.90  216541.120   

district_ar      منطي  ميت حلفا   ميت عقبة   ميت نما      هرم 2          هرم 3  
month                                                                           
4            59036.05  42099.55  148231.65  26359.20  778086.19  807098.591667  
5            50500.95  12225.10  153080.95  18467.99  697841.42  646077.960000  

[2 rows x 64 columns]

In [22]:
temp_df['total_monthly_revenue_growth'] = np.round(((total_monthly_revenue.loc[5,:] - total_monthly_revenue.loc[4,:]) / total_monthly_revenue.loc[4,:])*100,1)
total_revenue = temp_df.sort_values('total_monthly_revenue_growth', ascending=False)

data = px.treemap(data_frame= temp_df, 
                  path=[temp_df.index, 'total_monthly_revenue_growth'], 
                  color_continuous_scale='deep',
                  color='total_monthly_revenue_growth')

fig = go.Figure(data=data)
fig.update_layout(font=dict(family='Times New Roman', size=12), 
                  width=1200, 
                  height=800)
fig.show()

## Insights

## 1.  Top 5 Revenue Growth Districts

In [23]:
temp_df['total_monthly_revenue_growth'] = np.round(((total_monthly_revenue.loc[5,:] - total_monthly_revenue.loc[4,:]) / total_monthly_revenue.loc[4,:])*100,1)
total_revenue = temp_df.sort_values('total_monthly_revenue_growth', ascending=False)
total_revenue

number_of_orders_growth  distinct_retailers_growth  \
district_ar                                                           
شبين القناطر                       200.0                        0.0   
قلج                                 77.8                      100.0   
بلقس                                88.9                       42.9   
الخانكة                             55.6                      100.0   
القناطر الخيرية                     12.3                       -7.1   
...                                  ...                        ...   
الوايلي                            -45.9                      -32.1   
النزهة الجديدة                     -39.3                      -35.3   
ميت حلفا                           -40.0                        0.0   
مساكن شيراتون                      -66.7                      -37.5   
عمرانيه 3                          -75.0                      -63.6   

                 average_order_price_growth  total_monthly_revenue_growth  
district_ar                                                                
شبين القناطر                           10.4                         231.1  
قلج                                   -12.5                          55.6  
بلقس                                  -20.2                          50.8  
الخانكة                                -5.8                          46.5  
القناطر الخيرية                         9.3                          22.8  
...                                     ...                           ...  
الوايلي                               -13.4                         -53.1  
النزهة الجديدة                        -32.9                         -59.3  
ميت حلفا                              -51.6                         -71.0  
مساكن شيراتون                         -49.3                         -83.1  
عمرانيه 3                             -58.0                         -89.5  

[64 rows x 4 columns]

In [24]:
# Plot
top_5 = total_revenue.head()
least_5 = total_revenue.tail()

# Top 5 Bar Plot
fig1 = px.bar(data_frame=top_5, 
               x=top_5.index, 
               y='total_monthly_revenue_growth',
               text='total_monthly_revenue_growth',
               color='total_monthly_revenue_growth', 
               color_continuous_scale='deep')

fig1.update_xaxes(ticklabelposition="inside top", title='District')
fig1.update_yaxes(ticklabelposition="inside top", title='Revenue Growth (%)')
fig1.update_layout(font=dict(family='Arial', size=20))

fig1.show()

## 2.  Least 5 Revenue Growth Districts

In [25]:
# Least 5 Bar Plot
fig2 = px.bar(data_frame=least_5, 
              x=least_5.index, 
              y='total_monthly_revenue_growth', 
              text='total_monthly_revenue_growth',
              color='total_monthly_revenue_growth', 
              color_continuous_scale='deep')

fig2.update_xaxes(ticklabelposition="inside top", title='District')
fig2.update_yaxes(ticklabelposition="inside top", title='Revenue Growth (%)')
fig2.update_layout(font=dict(family='Arial', size=20))
fig2.show()

# Task 3 - Total Monthly Burn of April and May

In [26]:
# Calculate Burn rate for April and May
total_burn_april, total_burn_may = april_may_data.groupby(['month'])['discount'].sum()

# Plot Relation
fig = px.line(x=['April', 'May'], 
              y=[total_burn_april, total_burn_may], 
              markers=True)

fig.update_xaxes(title='Month')
fig.update_yaxes(title='Burn Rate')
fig.update_layout(font=dict(family='Arial', size=24), 
                  title= 'Burn Rate [ April - May ]', 
                  title_x = 0.5)
fig.show()

# Task 4 - Different Channel's Weekly Contribution [April - May]

In [27]:
# Building Pivot table
weekly_channel_contribution = pd.pivot_table(april_may_data, 
                                             values='sales_order_id', 
                                             index=['month', 'week'], 
                                             columns='channel' , 
                                             aggfunc='count' 
                                             )
weekly_channel_contribution

channel     Application  salesagent  telesales
month week                                    
4     14            557          53         71
      15           1600         126        242
      16           1907         142        227
      17           1891          81        191
      18           1308         147        118
5     18            165           4         15
      19           1660         418        170
      20           1694         372        180
      21           1771         250        189
      22            720          59         47
      23            450          44         49

In [28]:
# Drop missing values
weekly_channel_contribution.dropna(axis=1, inplace=True)

# Calculating Weekly Percentage of each channel
weekly_channel_contribution['Total'] = weekly_channel_contribution['Application'] +  weekly_channel_contribution['salesagent'] +  weekly_channel_contribution['telesales'] 
weekly_contribution_percentage = pd.DataFrame(columns=['application', 'salesagent','telesales'], 
                                              index=weekly_channel_contribution.index)
weekly_contribution_percentage['application'] = np.round((weekly_channel_contribution['Application'] / weekly_channel_contribution['Total'])*100,2)
weekly_contribution_percentage['salesagent'] = np.round((weekly_channel_contribution['salesagent'] / weekly_channel_contribution['Total'])*100,2)
weekly_contribution_percentage['telesales'] = np.round((weekly_channel_contribution['telesales'] / weekly_channel_contribution['Total'])*100,2)


# Plot 
fig = px.line(weekly_contribution_percentage, 
              x=weekly_contribution_percentage.index.droplevel(), 
              y=['application', 'salesagent', 'telesales'],
              markers=True)

fig.update_layout(title_text='Channel Weekly Percentage Contribution [April - May]', 
                  title_x=0.5, 
                  font=dict(family='Arial', size=20),
                  legend_title_text='Channel' 
                  )
fig.update_xaxes(ticklabelposition="inside top", title='Week')
fig.update_yaxes(ticklabelposition="inside top", title='Weekly Percentage Contribution (%)')
fig.show()

# Task 5 - Total Monthly Percentage of App Users & Telesales Users - May

## 1. Getting App Users Percentage

In [29]:
# Filtering Data of Delivered Orders
delivered_order_data = sales_data[sales_data['sales_order_status'] == 'Delivered'].reset_index(drop=True)

# Condition 1 - At least 2 months of delivered orders
temp=[]
two_month_delivered_ids = []
for index, row in delivered_order_data.iterrows():
    if row['retailer_id'] not in temp:
        temp.append(row['retailer_id'])
    else:
        two_month_delivered_ids.append(row['retailer_id'])

# check if retailer ID exists in two_month_delivered_ids
pd.set_option('mode.chained_assignment', None)
delivered_order_data['two_month_delivered'] = delivered_order_data['retailer_id'].isin(two_month_delivered_ids)
delivered_order_data

week  month creation_date  sales_order_id  retailer_id  \
0        14      4    2020-04-01               1        18564   
1        14      4    2020-04-01               5        28140   
2        14      4    2020-04-01              13        19299   
3        14      4    2020-04-01              17         5901   
4        14      4    2020-04-01              29        17783   
...     ...    ...           ...             ...          ...   
18354    25      6    2020-06-15           86868        28267   
18355    25      6    2020-06-15           86872        30883   
18356    25      6    2020-06-15           86880        16638   
18357    25      6    2020-06-15           86884        36652   
18358    25      6    2020-06-15           86888        39303   

      sales_order_status      channel district_ar  \
0              Delivered  Application   العمرانية   
1              Delivered  Application    ميت عقبة   
2              Delivered  Application    ميت عقبة   
3              Delivered  Application     المطرية   
4              Delivered  Application       هرم 3   
...                  ...          ...         ...   
18354          Delivered  Application      المقطم   
18355          Delivered  Application       الدقي   
18356          Delivered  Application        فيصل   
18357          Delivered  Application       مسطرد   
18358          Delivered  Application  دار السلام   

       order_price_before_discount  discount  revenue  two_month_delivered  
0                          2119.75      0.00  2119.75                 True  
1                          2242.75      0.10  2242.65                 True  
2                          3448.50     24.51  3423.99                 True  
3                           910.95      0.10   910.85                 True  
4                          2113.45      0.10  2113.35                 True  
...                            ...       ...      ...                  ...  
18354                      2502.50     50.05  2452.45                 True  
18355                      3309.00     46.52  3262.48                 True  
18356                      2294.20     39.76  2254.44                 True  
18357                      1660.35     18.92  1641.43                False  
18358                       966.60      0.10   966.50                 True  

[18359 rows x 12 columns]

In [30]:
# Filter data by Channels
app_delivered_order_data = delivered_order_data[delivered_order_data['channel'] == 'Application'].reset_index(drop=True)

# Application Delivered Orders Count
app_delivered_orders_per_retailer = pd.pivot_table(data=app_delivered_order_data, 
                                               index='retailer_id', 
                                               columns='month', 
                                               values='sales_order_id', 
                                               aggfunc='count')

app_delivered_orders_per_retailer.replace(np.nan, 0, inplace=True)
app_delivered_orders_per_retailer

month          4    5    6
retailer_id               
68           1.0  0.0  0.0
73           1.0  0.0  0.0
80           1.0  0.0  0.0
96           1.0  1.0  0.0
98           1.0  0.0  0.0
...          ...  ...  ...
40475        0.0  0.0  1.0
40571        0.0  0.0  1.0
40616        0.0  0.0  1.0
40657        0.0  0.0  1.0
40662        0.0  0.0  1.0

[6029 rows x 3 columns]

In [31]:
# Total Monthly Delivered Orders
total_delivered_orders_per_retailer = pd.pivot_table(data=delivered_order_data, 
                                               index='retailer_id', 
                                               columns='month', 
                                               values='sales_order_id', 
                                               aggfunc='count')

total_delivered_orders_per_retailer.replace(np.nan, 0, inplace=True)
total_delivered_orders_per_retailer

month          4    5    6
retailer_id               
68           1.0  0.0  0.0
73           1.0  0.0  0.0
80           1.0  0.0  0.0
85           1.0  1.0  0.0
90           1.0  0.0  0.0
...          ...  ...  ...
40742        0.0  0.0  1.0
40745        0.0  0.0  1.0
40753        0.0  0.0  1.0
40770        0.0  0.0  1.0
40778        0.0  0.0  1.0

[8356 rows x 3 columns]

In [32]:
# Getting Percentage of Application use for every retailer
app_monthly_delivery_percentage = np.round((app_delivered_orders_per_retailer / total_delivered_orders_per_retailer) * 100, 1)

app_monthly_delivery_percentage.dropna(subset=[5], inplace=True)

app_monthly_delivery_percentage

month            4      5      6
retailer_id                     
96           100.0  100.0    NaN
126          100.0  100.0    NaN
130            NaN  100.0    NaN
136            NaN  100.0    NaN
242          100.0  100.0    NaN
...            ...    ...    ...
38711          NaN  100.0    NaN
38737          NaN  100.0    NaN
38767          NaN    0.0  100.0
38768          NaN  100.0    NaN
38941          NaN  100.0  100.0

[3510 rows x 3 columns]

In [33]:
# Condition 2 for App User - 75% of their delivered monthly orders is from the App
above_75_ids = []
for index, row in app_monthly_delivery_percentage.iterrows():
    # Check if Application use percentage is higher than 75%
    if row.loc[5] > 75:
        above_75_ids.append(index)
    else:
        pass

# check if retailer ID exists in above_75_ids
pd.set_option('mode.chained_assignment', None)
delivered_order_data['is_above_75'] = delivered_order_data['retailer_id'].isin(above_75_ids)
delivered_order_data

week  month creation_date  sales_order_id  retailer_id  \
0        14      4    2020-04-01               1        18564   
1        14      4    2020-04-01               5        28140   
2        14      4    2020-04-01              13        19299   
3        14      4    2020-04-01              17         5901   
4        14      4    2020-04-01              29        17783   
...     ...    ...           ...             ...          ...   
18354    25      6    2020-06-15           86868        28267   
18355    25      6    2020-06-15           86872        30883   
18356    25      6    2020-06-15           86880        16638   
18357    25      6    2020-06-15           86884        36652   
18358    25      6    2020-06-15           86888        39303   

      sales_order_status      channel district_ar  \
0              Delivered  Application   العمرانية   
1              Delivered  Application    ميت عقبة   
2              Delivered  Application    ميت عقبة   
3              Delivered  Application     المطرية   
4              Delivered  Application       هرم 3   
...                  ...          ...         ...   
18354          Delivered  Application      المقطم   
18355          Delivered  Application       الدقي   
18356          Delivered  Application        فيصل   
18357          Delivered  Application       مسطرد   
18358          Delivered  Application  دار السلام   

       order_price_before_discount  discount  revenue  two_month_delivered  \
0                          2119.75      0.00  2119.75                 True   
1                          2242.75      0.10  2242.65                 True   
2                          3448.50     24.51  3423.99                 True   
3                           910.95      0.10   910.85                 True   
4                          2113.45      0.10  2113.35                 True   
...                            ...       ...      ...                  ...   
18354                      2502.50     50.05  2452.45                 True   
18355                      3309.00     46.52  3262.48                 True   
18356                      2294.20     39.76  2254.44                 True   
18357                      1660.35     18.92  1641.43                False   
18358                       966.60      0.10   966.50                 True   

       is_above_75  
0             True  
1            False  
2             True  
3             True  
4             True  
...            ...  
18354         True  
18355         True  
18356        False  
18357        False  
18358        False  

[18359 rows x 13 columns]

In [34]:
# Filter App users
app_users = delivered_order_data[delivered_order_data['two_month_delivered'] & delivered_order_data['is_above_75']].reset_index(drop=True)
may_app_users = app_users[app_users['month']==5].reset_index(drop=True)
may_app_users

week  month creation_date  sales_order_id  retailer_id  \
0       18      5    2020-05-01           34647        26658   
1       18      5    2020-05-01           34651         8913   
2       18      5    2020-05-01           34655        20805   
3       18      5    2020-05-01           34659        24977   
4       18      5    2020-05-01           34663         8544   
...    ...    ...           ...             ...          ...   
4180    23      5    2020-05-31           67660        18413   
4181    23      5    2020-05-31           67668        30589   
4182    23      5    2020-05-31           67672        24375   
4183    23      5    2020-05-31           67676        13651   
4184    23      5    2020-05-31           67680         5374   

     sales_order_status      channel    district_ar  \
0             Delivered  Application         المقطم   
1             Delivered  Application        المطرية   
2             Delivered  Application    باب الشعرية   
3             Delivered  Application        عين شمس   
4             Delivered  Application    حدائق القبة   
...                 ...          ...            ...   
4180          Delivered  Application         فيصل 3   
4181          Delivered  Application           فيصل   
4182          Delivered  Application        عين شمس   
4183          Delivered  Application  بولاق الدكرور   
4184          Delivered  Application         امبابة   

      order_price_before_discount  discount   revenue  two_month_delivered  \
0                         1359.20      3.59   1355.61                 True   
1                         3300.45     14.33   3286.12                 True   
2                        10947.15     39.51  10907.64                 True   
3                         3463.75     23.10   3440.65                 True   
4                         2981.75     29.07   2952.68                 True   
...                           ...       ...       ...                  ...   
4180                      2541.75     42.95   2498.80                 True   
4181                      1367.00      0.00   1367.00                 True   
4182                      2180.50     39.09   2141.41                 True   
4183                      3097.20     50.74   3046.46                 True   
4184                      3746.70     33.89   3712.81                 True   

      is_above_75  
0            True  
1            True  
2            True  
3            True  
4            True  
...           ...  
4180         True  
4181         True  
4182         True  
4183         True  
4184         True  

[4185 rows x 13 columns]

In [35]:
may_total_users = delivered_order_data[delivered_order_data['month'] == 5].reset_index(drop=True)

# Calculate Percentage of App Users in May
total_may_users_count = may_total_users.shape[0]
app_may_users_count = may_app_users.shape[0]
may_app_users_percentage = np.round((app_may_users_count / total_may_users_count) * 100, 0)
may_app_users_percentage

62.0

## 2. Filtering Telesales Users

In [36]:
#  = delivered_order_data[delivered_order_data['channel'] == 'telesales'].reset_index(drop=True)
# total_may_users_count

# Condition 1: 2 Month delivered Orders
telesales_delivered_order_data = delivered_order_data[delivered_order_data['two_month_delivered']].reset_index(drop=True)

# Condition 2: Last Delivered Order Channel telesales
telesales_delivered_order_data.groupby(['retailer_id', 'channel'], as_index=False).last()
telesales_users = telesales_delivered_order_data[telesales_delivered_order_data['channel'] == 'telesales'].reset_index(drop=True)

# Getting May Telesales Users
may_telesales_users = telesales_users[telesales_users['month'] == 5].reset_index(drop=True)
may_telesales_users

week  month creation_date  sales_order_id  retailer_id  \
0      18      5    2020-05-01           34815         6961   
1      18      5    2020-05-01           34827        22763   
2      18      5    2020-05-01           34899        27341   
3      18      5    2020-05-01           35019        15611   
4      18      5    2020-05-01           35023         6262   
..    ...    ...           ...             ...          ...   
480    23      5    2020-05-31           67104        20400   
481    23      5    2020-05-31           67220         9826   
482    23      5    2020-05-31           67288         8063   
483    23      5    2020-05-31           67500        22280   
484    23      5    2020-05-31           67664        17123   

    sales_order_status    channel      district_ar  \
0            Delivered  telesales           امبابة   
1            Delivered  telesales      شبرا الخيمة   
2            Delivered  telesales            قليوب   
3            Delivered  telesales             فيصل   
4            Delivered  telesales           الوراق   
..                 ...        ...              ...   
480          Delivered  telesales         العباسية   
481          Delivered  telesales  الزاوية الحمراء   
482          Delivered  telesales      شبرا الخيمة   
483          Delivered  telesales            المرج   
484          Delivered  telesales            هرم 2   

     order_price_before_discount  discount  revenue  two_month_delivered  \
0                        2013.00     27.78  1985.22                 True   
1                        4017.10     58.95  3958.15                 True   
2                        1787.50      0.00  1787.50                 True   
3                        2951.30     16.43  2934.87                 True   
4                        8833.75    114.30  8719.45                 True   
..                           ...       ...      ...                  ...   
480                      1149.00     15.03  1133.97                 True   
481                      2472.50     29.69  2442.81                 True   
482                       922.20      8.82   913.38                 True   
483                      1915.50     16.54  1898.96                 True   
484                      1922.25     12.82  1909.43                 True   

     is_above_75  
0          False  
1          False  
2          False  
3          False  
4          False  
..           ...  
480        False  
481        False  
482        False  
483        False  
484        False  

[485 rows x 13 columns]

In [37]:
# Calculate Percentage of Telesales Users in May
telesales_may_users_count = may_telesales_users.shape[0]
may_telesales_users_percentage = np.round((telesales_may_users_count / total_may_users_count) * 100, 0)
may_telesales_users_percentage

7.0

In [38]:
# Plot the results
fig =  px.bar(x=['Application Users', 'Telesales Users'], 
              y=[may_app_users_percentage, may_telesales_users_percentage],
              text=[may_app_users_percentage, may_telesales_users_percentage])

fig.update_xaxes(title='')
fig.update_yaxes(ticklabelposition="inside top", title='Percentage Use (%)')
fig.update_layout(font=dict(family='Arial', size=24), 
                  title='App Users Vs Telesales Users Percentage [May]',
                  title_x = 0.5)
fig.update_traces(marker_color=['#DC3912','#00B5F5'])
fig.show()

# Task 6 - Projection Of Total Sales Per Retailer in June

In [39]:
'''
Calculate the projection of total sales orders and total revenues (sum of order price) 
for the rest of June for every retailer and hence, total June.
(use the assumptions that you find reasonable in your answer)

'''
sales_per_retailer = pd.pivot_table(data=sales_data, 
                                    index='retailer_id', 
                                    columns='month', 
                                    values='order_price_before_discount', 
                                    aggfunc='sum')

sales_per_retailer.replace(np.nan,0, inplace=True)
sales_per_retailer

month              4        5        6
retailer_id                           
68           1055.30     0.00     0.00
73           1783.20     0.00     0.00
80            855.50     0.00     0.00
85           6793.60  9966.25     0.00
90           1382.85     0.00     0.00
...              ...      ...      ...
40745           0.00     0.00   823.25
40753           0.00     0.00   847.70
40758           0.00     0.00   933.30
40770           0.00     0.00   932.00
40778           0.00     0.00  1671.00

[9276 rows x 3 columns]

In [40]:
# June Retailers = Retailers who have sales in June or do not have sales in both April and May
june_retailers = sales_per_retailer[((sales_per_retailer[4] == 0) & (sales_per_retailer[5] == 0)) | (sales_per_retailer[6] != 0)]
june_retailers

month             4       5         6
retailer_id                          
98           2471.0     0.0   3636.50
120          3150.6     0.0   1199.05
129             0.0  9326.8  13049.75
209             0.0     0.0   3051.00
217             0.0     0.0    830.75
...             ...     ...       ...
40745           0.0     0.0    823.25
40753           0.0     0.0    847.70
40758           0.0     0.0    933.30
40770           0.0     0.0    932.00
40778           0.0     0.0   1671.00

[3681 rows x 3 columns]

In [41]:
sales_per_retailer.drop(index= june_retailers.index, inplace=True)
sales_per_retailer

month              4        5    6
retailer_id                       
68           1055.30     0.00  0.0
73           1783.20     0.00  0.0
80            855.50     0.00  0.0
85           6793.60  9966.25  0.0
90           1382.85     0.00  0.0
...              ...      ...  ...
38946           0.00  2330.15  0.0
38952           0.00  1552.75  0.0
38983           0.00  2219.00  0.0
38986           0.00  6151.50  0.0
39003           0.00  2413.00  0.0

[5595 rows x 3 columns]

In [42]:
sales_per_retailer[6] = (sales_per_retailer[4] + sales_per_retailer[5]) / 2
sales_per_retailer

month              4        5         6
retailer_id                            
68           1055.30     0.00   527.650
73           1783.20     0.00   891.600
80            855.50     0.00   427.750
85           6793.60  9966.25  8379.925
90           1382.85     0.00   691.425
...              ...      ...       ...
38946           0.00  2330.15  1165.075
38952           0.00  1552.75   776.375
38983           0.00  2219.00  1109.500
38986           0.00  6151.50  3075.750
39003           0.00  2413.00  1206.500

[5595 rows x 3 columns]

## Predicted Total Sales Order Per Retailer In June

In [43]:
temp = pd.concat([june_retailers,sales_per_retailer], axis=0)
june_predicted_sales_per_retailer = temp[6].to_frame()
june_predicted_sales_per_retailer.sort_values(6, ascending=False, inplace=True)
june_predicted_sales_per_retailer.rename(columns={6:'june_sales'}, inplace=True)
june_predicted_sales_per_retailer

june_sales
retailer_id            
24751         98316.000
26549         92909.000
19921         68314.700
25782         63079.575
7819          62827.450
...                 ...
16861             0.000
29678             0.000
31969             0.000
33360             0.000
26158             0.000

[9276 rows x 1 columns]

## Visualizing Top 10 Retailer IDs in terms of Revenues in June

In [44]:
top_10_retailers_june = june_predicted_sales_per_retailer.head(10)
top_10_retailers_june.reset_index(inplace=True)
top_10_retailers_june['retailer_id'] = top_10_retailers_june['retailer_id'].astype('str')
top_10_retailers_june
# Plot
fig = px.bar(data_frame=top_10_retailers_june, 
              x='retailer_id', 
              y='june_sales', 
              text='june_sales',
              color='june_sales', 
              color_continuous_scale='deep')

fig.update_xaxes(ticklabelposition="inside top", title='Retailer ID')
fig.update_yaxes(ticklabelposition="inside top", title='Total Revenue')
fig.update_layout(font=dict(family='Arial', size=20), 
                  title="Top 10 Revenue Retailer IDs [June]", 
                  title_x=0.5)
fig.show()

## Visualizing Total Monthly Revenue [April - June]

In [45]:
sales_per_retailer.reset_index(inplace=True, drop=True)
# sales_per_retailer.drop('retailer_id', inplace=True)
total_monthly_sales = sales_per_retailer.sum(axis=0).to_frame()
total_monthly_sales.rename(columns={0:'total_monthly_sales'}, inplace=True)
total_monthly_sales

total_monthly_sales
month                     
4             1.544231e+07
5             1.194461e+07
6             1.369346e+07

In [46]:
# Plot
fig = px.line(data_frame= total_monthly_sales, 
             x=['April', 'May', 'June'], 
             y='total_monthly_sales', 
              markers=True)

fig.update_layout(title_text='Total Monthly Revenue [April - June]', 
                  title_x=0.5, 
                  font=dict(family='Arial', size=20),
                  legend_title_text='Channel' 
                  )

fig.update_xaxes(title=" ")
fig.update_yaxes(ticklabelposition="inside top", title='Total revenue')
fig.show()